keps/sig-auth/20190926-dynamic-audit-policy.md

```diff
@@ -0,0 +1,387 @@
+
+Understanding stages requires deep understanding of the various request processing options and details. The stages are not strictly sequential. For example, request processing enters the `ResponseStarted` stage only for long running operations, which applies only to request with verbs `watch` and/or `proxy` and/or urls such as `pprof` and/or other conditions that are not externalized.
+ 
+Using certain stages with level combination could produce cumbersome results. For example, switching off emit of audit events to a sink could be achieved both by defining `level: None`, or by not providing `stages`. It was easy to produce incoherent, contradicting presets such as `level` that is different from `None` and no `stages` or `stages` set and `level` set to `None`.
+
+The proposed API takes a simpler and less error-prone approach and treats stages strictly *internal*. In contrast to the file policy format, stages are specified neither in policies, nor in rules. Instead, every rule is *implicitly* assigned a whitelist of stages (`ResponseStarted`, `ResponseComplete`, `Panic`), excluding `RequestReceived`. 
+
+`RequestReceived` is generally not applicable to dynamic audit webhook backends, because they are designed for *non-blocking* (batch) scenarios. `RequestReceived` was added to guarantee an audit event was logged even in a "query of doom" scenario. However, that guarantee only holds if auditing is in *synchronous* mode to ensure the query of doom is not processed until the event is already successfully recorded. In batch mode, the `RequestReceived` event is put in the buffer, but when the API server is taken offline it is never delivered.
+
+The effective stage, for which an audit event was generated is provided by the [`Stage`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go#L81) property of the `Event` type. 
+
+#### Request Selection Rules
+
+> For a more comprehensive referece to rules concept support, please refer to the [Dynamic Audit Policy - Design Proposal](https://docs.google.com/document/d/1Mp1E4fEbFiCBmSrAJxZuPrbwjYMZ4dAZC0K4XvSyZFA/edit?usp=sharing#heading=h.i02gvql48q0x).
+
+Rules defined by `AuditClass` objects are used to select requests for audit level assignment in an `AuditSink` policy. There must be at least one rule, or an error is yielded. Generally, the order of rules in an `AuditClass` is not important as long as at least one rule matches a request. However, accounting that rules are matched on the *first-matches-wins* principle for reconciliation among potentially multiple overlapping rules, will lead to more concise manifests.
+
+Requests matched by rule are *selected*. Rules define the criteria to match requests. The way it is interpreted is no different to the file policies.
+
+Selection rules are typed:
```

Comment:
Please, state your concerns.

The design driver was to make them as explicit as possible. Having supplementary documentation on what is _actually_ selectors combination and values valid is not exactly dev-friendly. I'm open to suggestions how to achieve that in other (better?) ways.

Reply:
My concerns are mostly around useability, since the approaches we're debating are equivalent in expressiveness.

If I want to include resources X, Y, and Z and don't care about what namespaces they're in, I need to know whether each one is cluster-scoped or namespace scoped. This is also error prone, as I don't think we can validate that the real resources have the same scope as that listed here. I also think there is some redundancy in the current approach.

Thinking about this a bit more, I think my main concern is around separating cluster & namespace scoped resources. Here's an alternative approach (strawman):

```go
type GroupResourceSelector struct { 
  // +optional
  Group string
  // +optional
  Resources []ResourceSelector
  // +optional
  Scope ScopeSelector
}

type ResourceSelector struct {
  Kind string
  // Subresources on this Kind.
  // nil = no subresources.
  // '*' = all subresources & non-subresourced
  // '' = non-subresourced
  // +optional
  Subresources []string
  // +optional
  ObjectNames []string
}

// The scope selector makes the cluster vs. namespaced scope more explicit
// +union
type ScopeSelector struct {
  // +unionDiscriminator
  Scope ScopeType
  // +optional
  Namespaces []string
}

const (
  ScopeTypeAny = "Any" // default
  ScopeTypeCluster = "Cluster"
  ScopeTypeNamespaced = "Namespaced"
)
```

This still feels too clunky to me. I'll keep thinking about it...

Reply:
I see where you're going. I like the idea of merging namespace into the scope of GroupResource(Selector) structure. That's where it belongs indeed.   
I have two concerns though (or maybe one if redundancy is not a thing anymore):
1. With the proposed approach you can still do the contradiction:
    ```
    scope: 
      scope: Cluster
      namespaces: ["x","y","z"]
    ```
    , which is like the following possible error you have to keep an eye on today in file policies:
    ```
    - group: "" 
      resources: ["configmaps"]
      namespaces: [""]
    ```
    On the bright side, I agree it's more explicit in your proposal than it was in file policies and it does not build on the special semantics of empty namespace string in a namespaces array. That's still an improvement over file policy.
2. You can't avoid `verbs` and `users` selectors redundancy even in this proposal. As long as we do any kind of breakdown into selector types, they have to be repeated in every single one because they can _always_ be part of the criteria.

Reply:
Reading through abut unions this starts to shape for me too. Anyway a couple of questions:
- What does it mean if `Scope` in `GroupResourceSelector` is `undefined`? Same as `Any` scope?
- How does scope type `Any` translate to the current policy model? I can generate a cluster resource rule. But I can't generate an _any_ namespace rue, correct?

Reply:
Let's recap and drive this to the finish line.

The original proposal in this KEP follows the [RBAC approach](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole) of distinction between cluster and namespaced resources with dedicated selector struct types. Unlike it, it distinguishes non resource request selectors as a distinct selector type, and models the "any" namespace selection in the resource selector type rather than in the cluster resource selectors type.

The proposed change borrows the `scope` concept from the [Dynamic admission control approach](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules) for request selection rules reducing the two types of resource selection structs to one, using a discriminator for the resource scope (`cluster`|`namespaced`|`any`). In addition it introduces a more explicit and fine-grained selection model for resources (incl. explicitly `subresources` now)

The example from the KEP implemented as proposed here, would look like this:
```yaml
apiVersion: auditregistration.k8s.io/v1alpha1
metadata:
  name: sensitive-things 
kind: AuditClass
spec:
  rules:
  - userGroups: ["system:masters"]
    verbs: ["create","patch","update","delete"]
    groupResourceSelectors: # match secrets and configmaps in the kube-system namespace
    - group: ""
      resources: 
      - kind: secrets
      - kind: configmaps
      scope: Namespaced
      namespaces: ["kube-system"]	  
  - groupResourceSelectors: # match auditclass cluster resources
    - group: "auditregistration.k8s.io"
      resources: 
      - kind: auditclasses
      scope: Cluster
---
apiVersion: auditregistration.k8s.io/v1alpha1
kind: AuditClass
metadata:
  name: noisy-lowrisk-things
spec:
  rules:
  - groupResourceSelectors:  # match secrets and configmaps with name controller-leader in the kube-system namespace
    - group: ""
      resources: 
      - kind: configmaps
        objectNames: ["controller-leader"]
      scope: Namespaced
      namespaces: ["kube-system"]
  - users: ["system:kube-proxy"]
    verbs: ["watch"]	  
    - groupResourceSelectors: # match services and endpoints cluster-wide
      - group: ""	
        resources: 
        - kind: services
        - kind: endpoints
  - userGroups:  ["system:authenticated"]
    - nonResourceSelectors:  # match requests to URL paths patterns  "/api*" or "/version"
      urls: ["/api*","/version"]
```
It seems we [won't be able to](https://github.com/kubernetes/enhancements/issues/1027#issuecomment-534381561) get advantage of the [unions](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md) enhancement for now, but the foundations for adopting it later on will be laid down.

Finally, I would like to propose one more improvement concerning selection of _generated_ or _high volume_ objects. 
> I have concerns about 1.17 timelines, so on this stage the minimum that I want is to confirm it and ensure that it can be introduced with a subsequent API update. For that I will need the `namespaces` list components to be composite objects.

Generated and high volume objects are a common case for example for providers of managed environments. Think of namespaces for consumers control planes. In this use case the lack of convenient way to select a potentially enormous amount of objects (e.g. namespaces) is a bigger concern than somebody applying or removing a label, since the cluster control is concentrated and controlled anyway. 

Also considering that one of the major scenarios is debugging, there is higher demand for flexibility in selecting audited objects, incl. generated ones. For example, you'd be interested in what is it that does funny things to your "my-app" pods for a deployment, not in all pods in a namespace in general, esp. in large solutions. Of course that would be encouraged only in clusters where it is safe to rely on the objects labels.

I would like to change the `namespaces` selector to comprise composite structs allowing dynamic selection based on labels similar to what we have in dynamic admission control  [here](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-objectselector
) and [here](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-namespaceselector).

Considering the example above a selectors would then look like that:
```yaml
groupResourceSelectors: # match secrets and configmaps in the kube-system namespace
- group: ""
  resources: 
  - kind: secrets
  - kind: configmaps
  scope: Namespaced
  namespaces: 
  - name: kube-system
```
or using a namespace selector as proposed:
```yaml
groupResourceSelectors: # match secrets and configmaps in namespaces labeled as "environment:prod" or "environment:staging"
- group: ""
  resources: 
  - kind: secrets
  - kind: configmaps
  scope: Namespaced
  namespaces: 
  - matchExpressions:
    - key: environment
      operator: In
      values: ["prod","staging"]
```
Both elements `name` and `matchExpression` are mutually exclusive, i.e. a union modeling `oneOf` will be used.

Similarly,  a `matchExpression` could be added to `ResourceSelector` to achieve the same on object level like [here](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-objectselector
).